prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -5467,67 +5496,226 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         &mut self,
         local: mir::Local,
         projection: &[mir::PlaceElem<'tcx>],
+        encode_kind: ArrayAccessKind,
     ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>, ty::Ty<'tcx>, Option<usize>)> {
         let (encoded_place, ty, variant_idx) = self.mir_encoder.encode_projection(local, projection)?;
-        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place)?;
+        let (encoded_expr, encoding_stmts) = self.postprocess_place_encoding(encoded_place, encode_kind)?;
 
         Ok((encoded_expr, encoding_stmts, ty, variant_idx))
     }
 
+    fn encode_array_lookup_shared(
+        &mut self,
+        base: PlaceEncoding<'tcx>,
+        index: vir::Expr,
+        array_ty: ty::Ty<'tcx>,
+    ) -> EncodingResult<(vir::Expr, Vec<vir::Stmt>)> {
+        let (array_pred, array_ty, elem_pred, elem_ty, elem_value_ty, elem_ty_rs, array_len) =
+            self.encode_array_types(array_ty)?;
+
+        let lookup_pure = self.encoder.encode_builtin_function_use(
+            BuiltinFunctionKind::ArrayLookupPure {
+                array_ty_pred: array_pred.clone(),
+                elem_ty_pred: elem_pred,
+                array_len,
+                return_ty: elem_value_ty.clone(),
+            }
+        );
+
+        let lookup_res: vir::Expr = self.cfg_method.add_fresh_local_var(elem_ty).into();
+        let val_field = self.encoder.encode_value_field(elem_ty_rs);
+        let lookup_res_val_field = lookup_res.clone().field(val_field);
+
+        let (encoded_base_expr, mut stmts) = self.postprocess_place_encoding(base, ArrayAccessKind::Shared)?;
+        stmts.extend(self.encode_havoc_and_allocation(&lookup_res));
+
+        let usize_ty = self.encoder.env().tcx().mk_ty(ty::TyKind::Uint(ty::UintTy::Usize));
+        let idx_val_int = index.field(self.encoder.encode_value_field(usize_ty));
+
+        let lookup_pure_call = vir::Expr::func_app(
+            lookup_pure,
+            vec![
+                encoded_base_expr.clone(),
+                idx_val_int.clone()
+            ],
+            vec![
+                vir::LocalVar::new(
+                    String::from("self"),
+                    array_ty.clone(),
+                ),
+                vir::LocalVar::new(
+                    String::from("idx"),
+                    vir::Type::Int,
+                ),
+            ],
+            elem_value_ty,
+            vir::Position::default(),
+        );
+
+        stmts.push(vir::Stmt::Assert(
+            vir!{ [idx_val_int.clone()] < [ vir::Expr::from(array_len) ] },
+            vir::Position::default(),
+        ));
+        stmts.push(vir::Stmt::Assert(
+            vir::Expr::predicate_access_predicate(array_pred, encoded_base_expr, vir::PermAmount::Read),
+            vir::Position::default(),
+        ));
```

Comment:
Why is this assertion generated? It seems redundant to me.

Reply:
good catch, thanks, will adjust :-)